### PR TITLE
Add function and interpolation mode to style property docs

### PR DESCRIPTION
**Pull request by boundsj**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/7935

This adds documentation to each style layer property that specifies the function and interpolation mode combinations that will work for the property. 

In some cases, function and interpolation stop combination values that don't make sense will be coerced into a values that do. For example, a camera function with and exponential interpolation mode that is set on a property that is a piecewise-constant function will be converted into a camera function with an interval interpolation mode.

### CODE REVIEW

File: platform/darwin/scripts/generate-style-code.js [hunk 1]
```diff
@@ -281,6 +281,18 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
             doc += `\n\nThis attribute corresponds to the <a href="https://www.mapbox.com/mapbox-gl-style-spec/#${anchor}"><code>${property.original}</code></a> layout property in the Mapbox Style Specification.`;
         }
     }
+    if (property["property-function"]) {
+        doc += '`\n\nThis property can be set to an `MGLStyleConstantValue`, an `MGLCameraStyleFunction` using an `MGLInterpolationModeExponential` or `MGLInterpolationModeInterval` ' +
+               'interpolation mode, an `MGLSourceStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval`, `MGLInterpolationModeCategorical, or `MGLInterpolationModeIdentity` ' + 
+               'interpolation mode, or an `MGLCompositeStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval` or `MGLInterpolationModeCategorical` interpolation mode`';
+    } else {
+        if (property.function == "interpolated") {
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: Nit: use the `===` operator unless type coercion is required.

File: platform/darwin/scripts/generate-style-code.js [hunk 2]
```diff
@@ -281,6 +281,18 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
...
+        doc += '`\n\nThis property can be set to an `MGLStyleConstantValue` or an `MGLCameraStyleFunction` using an `MGLInterpolationModeExponential` or `MGLInterpolationModeInterval` ' +
+               'interpolation mode';
+        } else {
+        doc += '`\n\nThis property can be set to an `MGLStyleConstantValue` or an `MGLCameraStyleFunction` using an `MGLInterpolationModeInterval` interpolation mode';
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: Funky indentation.

File: platform/darwin/scripts/generate-style-code.js [hunk 3]
```diff
@@ -281,6 +281,18 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
             doc += `\n\nThis attribute corresponds to the <a href="https://www.mapbox.com/mapbox-gl-style-spec/#${anchor}"><code>${property.original}</code></a> layout property in the Mapbox Style Specification.`;
         }
     }
+    if (property["property-function"]) {
+        doc += '`\n\nThis property can be set to an `MGLStyleConstantValue`, an `MGLCameraStyleFunction` using an `MGLInterpolationModeExponential` or `MGLInterpolationModeInterval` ' +
+               'interpolation mode, an `MGLSourceStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval`, `MGLInterpolationModeCategorical, or `MGLInterpolationModeIdentity` ' + 
+               'interpolation mode, or an `MGLCompositeStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval` or `MGLInterpolationModeCategorical` interpolation mode`';
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: End this sentence with a period.

File: platform/darwin/scripts/generate-style-code.js [hunk 4]
```diff
@@ -281,6 +281,18 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
             doc += `\n\nThis attribute corresponds to the <a href="https://www.mapbox.com/mapbox-gl-style-spec/#${anchor}"><code>${property.original}</code></a> layout property in the Mapbox Style Specification.`;
         }
     }
+    if (property["property-function"]) {
+        doc += '`\n\nThis property can be set to an `MGLStyleConstantValue`, an `MGLCameraStyleFunction` using an `MGLInterpolationModeExponential` or `MGLInterpolationModeInterval` ' +
+               'interpolation mode, an `MGLSourceStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval`, `MGLInterpolationModeCategorical, or `MGLInterpolationModeIdentity` ' + 
+               'interpolation mode, or an `MGLCompositeStyleFunction` using an `MGLInterpolationModeExponential`, `MGLInterpolationModeInterval` or `MGLInterpolationModeCategorical` interpolation mode`.';
```
Comment (friedbunny) [inline platform/darwin/scripts/generate-style-code.js]: The end of this string has a stray tick before the period.

Comment (friedbunny) [inline platform/darwin/scripts/generate-style-code.js]: Which leads to this, on `MGLCircleStyleLayer.circleOpacity`:

screen shot 2017-02-08 at 2 08 10 pm

**Comment by 1ec5** [comment]
> Concatenating strings is leading to long, difficult to follow sentences — could we organize these into tables or otherwise section them?

Yes. At the very least, we should use semicolons to separate list items that contain commas.

**Comment by boundsj** [comment]
@friedbunny @1ec5 I did not find markup for creating arbitrary tables in jazzy. Are the changes in a749e82 suitable? Example:

screenshot 2017-02-08 12 12 12

**Comment by friedbunny** [comment]
That looks much better, thanks @boundsj. For tables in jazzy, we’ve been resorting to HTML (but that tends to look terrible inline).

**Comment by friedbunny** [comment]
Pushed 9ae227e, which adds _even more bullets_:

screen shot 2017-02-08 at 6 00 36 pm

**Comment by incanus** [comment]
> _even more bullets_

**Comment by 1ec5** [comment]
> For tables in jazzy, we’ve been resorting to HTML (but that tends to look terrible inline).

We also have some instances of [Markdown tables](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/darwin/docs/guides/Working%20with%20GeoJSON%20Data.md) in the jazzy guides, but we only get away with that because the guides don’t appear in Quick Help or any headers.

**Comment by boundsj** [comment]
I'm liking the _even more bullets_ look anyway.

File: platform/darwin/scripts/generate-style-code.js [hunk 5]
```diff
@@ -281,6 +281,30 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
             doc += `\n\nThis attribute corresponds to the <a href="https://www.mapbox.com/mapbox-gl-style-spec/#${anchor}"><code>${property.original}</code></a> layout property in the Mapbox Style Specification.`;
         }
     }
+    doc += '\n\nThis property can be set to one of the following values:\n\n' +
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: We’re saying what kind of values the property can be set to, not which values specifically. Also, avoid passive voice:

> You can set this property to an instance of:

File: platform/darwin/scripts/generate-style-code.js [hunk 6]
```diff
@@ -281,6 +281,30 @@ global.propertyDoc = function (propertyName, property, layerType, kind) {
...
+           '- `MGLStyleConstantValue`\n';
+    if (property["property-function"]) {
+        doc += '- `MGLCameraStyleFunction` with an interpolation mode of:\n' +
```
Comment (1ec5) [inline platform/darwin/scripts/generate-style-code.js]: FYI, we’ve been using `*` syntax in other places, but I don’t think it matters much.

File: platform/darwin/src/MGLCircleStyleLayer.h [hunk 7]
```diff
@@ -9,34 +9,70 @@ NS_ASSUME_NONNULL_BEGIN
 
 /**
  Controls the scaling behavior of the circle when the map is pitched.
+ 
+ This property can be set to one of the following values:
```
Comment (1ec5) [inline platform/darwin/src/MGLCircleStyleLayer.h]: This isn’t a property, it’s an enumeration value. Enumeration values don’t need this extra text.

File: platform/darwin/scripts/generate-style-code.js (hunk 5 above)
Comment (boundsj) [inline platform/darwin/scripts/generate-style-code.js]: Fixed in c602839

File: platform/darwin/scripts/generate-style-code.js (hunk 6 above)
Comment (boundsj) [inline platform/darwin/scripts/generate-style-code.js]: Updated in c602839 for consistency

File: platform/darwin/src/MGLCircleStyleLayer.h (hunk 7 above)
Comment (boundsj) [inline platform/darwin/src/MGLCircleStyleLayer.h]: Thanks for catching this! Updated in c602839

**Pull request merged**